README
This is the Speech Dispatcher driver for Gnome Speech
(gnome-speech-speechd-driver). It is a part of the Free(b)soft project
(http://www.freebsoft.org/), which is intended to allow blind and visually
impaired people to work with computer and Internet based on free software.

To build and install this driver into Gnome Speech, read the file INSTALL.

For information about how to install and use Speech Dispatcher, see
	http:/www.freebsoft.org/speechd

For information about how to install and use Gnome Speech, see
	http://developer.gnome.org/projects/gap/	

Complete documentation may be found in doc directory. Read doc/README for more
information.

*******************************************************************************

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU Library General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your option) any
later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU Library General Public License for more
details (file COPYING in the root directory).

You should have received a copy of the GNU Library General Public License along
with this program; if not, write to the Free Software Foundation, Inc., 675
Mass Ave, Cambridge, MA 02139, USA.

*******************************************************************************


NEWS

Version 0.1
	* Gnome Speech Speech Dispatcher Synthesis Driver provides
	basic functionality so that Gnome Speech (and Gnopernicus)
	could synthesize it's speech output through Speech Dispatcher.

build.sh
#!/bin/bash

# build.sh - Run all the auto-tools
#
# Copyright (C) 2001,2002,2003 Brailcom, o.p.s, Prague 2,
# Vysehradska 3/255, 128 00, <freesoft@freesoft.cz>
#
# This is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this package; see the file COPYING.  If not, write to
# the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
# Boston, MA 02111-1307, USA.
#
# $Id: build.sh,v 1.1 2005-12-16 16:29:52 hanke Exp $


echo "Copying libtool macros (libtoolize)"
if ! libtoolize --force; then
	echo "libtoolize failed!"
	exit 1
fi

echo "Building user-defined autoconf macros (aclocal)"
if ! aclocal; then
	echo "aclocal failed!"
	exit 1
fi

echo "Creating ./configure (autoconf)"
if ! autoconf; then
	echo "autoconf failed!"
	exit 1
fi

echo "Creating config.h.in (autoheader)"
if ! autoheader; then
	echo "autoheader failed!"
	exit 1
fi

echo "Checking for missing scripts (automake -a)"
if ! automake -a; then
	echo "automake -a failed!"
	exit 1
fi

echo "Creating makefiles.in (automake)"
if ! automake; then
	echo "automake failed!"
	exit 1
fi

echo 
echo "You can continue configuring and compiling Speech Dispatcher with"
echo "       ./configure && make all && make install"


libspeechd.c

/*
 * libspeechd.c - Functions for acces to Speech Dispatcher functions
 *
 * Copyright (C) 2005 Brailcom, o.p.s.
 *
 * This is free software; you can redistribute it and/or modify it under the
 * terms of the GNU Library General Public License as published by the Free
 * Software Foundation; either version 2, or (at your option) any later
 * version.
 *
 * This software is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Library General Public License
 * along with this package; see the file COPYING.  If not, write to the Free
 * Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 *
 * $Id: libspeechd.c,v 1.1 2005-12-16 16:29:48 hanke Exp $
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/tcp.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <glib.h>
#include <errno.h>
#include <assert.h>
#include <ctype.h>

#include "libspeechd.h"

#ifdef SPD_DEBUG
#define DBG(arg...) \
{ \
    fprintf(stderr," SpeechdOutputLibrary: "); \
    fprintf(stderr,arg); \
    fprintf(stderr,"\n"); \
    fflush(stderr); \
}
#else
#define DBG(arg...) 
#endif

/* --------------------- Public functions ------------------------- */

/* Opens a new Speech Dispatcher connection.
 * Returns socket file descriptor of the created connection
 * or -1 if no connection was opened. */
int 
spd_open(const char* client_name, const char* connection_name, const char* user_name)
{
  struct sockaddr_in address;
  int connection;
  char *set_client_name;
  char* conn_name;
  char* usr_name;
  char *env_port;
  int port;
  int ret;
  char tcp_no_delay = 1;

  if (client_name == NULL) return -1;

  usr_name = strdup(user_name);
  
  if(connection_name == NULL)
      conn_name = strdup("main");
  else
      conn_name = strdup(connection_name);

  env_port = getenv("SPEECHD_PORT");
  if (env_port != NULL)
      port = strtol(env_port, NULL, 10);
  else
      port = SPEECHD_DEFAULT_PORT;
  
  /* Prepare a new socket */
  address.sin_addr.s_addr = inet_addr("127.0.0.1");
  address.sin_port = htons(port);
  address.sin_family = AF_INET;
  connection = socket(AF_INET, SOCK_STREAM, 0);

  /* Connect to server */
  ret = connect(connection, (struct sockaddr *)&address, sizeof(address));
  if (ret == -1){
      DBG("Error: Can't connect to server: %s", strerror(errno));
      return 0;
  }

  setsockopt(connection, IPPROTO_TCP, TCP_NODELAY, &tcp_no_delay, sizeof(int));
 
  set_client_name = g_strdup_printf("SET SELF CLIENT_NAME \"%s:%s:%s\"", usr_name,
          client_name, conn_name);

  ret = spd_execute_command(connection, set_client_name);

  spd__free(usr_name);
  spd__free(conn_name);
  spd__free(set_client_name);

  return connection;
}


/* Close a Speech Dispatcher connection */
void
spd_close(int connection)
{
    spd_execute_command(connection, "QUIT");
    
    /* close the socket */
    close(connection);
}

/* Say TEXT with priority PRIORITY.
 * Returns 0 on success, -1 otherwise. */                            
int
spd_say(int connection, SPDPriority priority, const char* text)
{
    static char command[16];
    char *etext;
    int ret;

    if (text == NULL) return -1;

    /* Set priority */
    ret = spd__set_priority(connection, priority);
    if(ret) return -1;
    
    /* Check if there is no escape sequence in the text */
    etext = spd__escape_dot(text);
    if (etext == NULL) etext = (char*) text;
  
    /* Start the data flow */
    sprintf(command, "SPEAK");
    ret = spd_execute_command(connection, command);
    if(ret){     
        DBG("Error: Can't start data flow!");
        return -1;
    }
  
    /* Send data */
    spd_send_data(connection, etext, SPD_NO_REPLY);

    /* Terminate data flow */
    ret = spd_execute_command(connection, "\r\n.");
    if(ret){
        DBG("Can't terminate data flow");
        return -1; 
    }

    return 0;
}

int
spd_stop(int connection)
{
  return spd_execute_command(connection, "STOP SELF");
}

int
spd_cancel(int connection)
{
  return spd_execute_command(connection, "CANCEL SELF");
}


int
spd_pause(int connection)
{
  return spd_execute_command(connection, "PAUSE SELF");
}

int
spd_resume(int connection)
{
  return spd_execute_command(connection, "RESUME SELF");
}

int
spd_sound_icon(int connection, SPDPriority priority, const char *icon_name)
{
    char *command;
    int ret;

    if (icon_name == NULL) return -1;

    ret = spd__set_priority(connection, priority);
    if (ret) return -1;

    command = g_strdup_printf("SOUND_ICON %s", icon_name);
    ret = spd_execute_command(connection, command);
    spd__free (command);
    if (ret) return -1;
    
    return 0;
}

#define SPD_SET_COMMAND_INT(param, ssip_name, condition) \
    int \
    spd_set_ ## param (int connection, signed int val) \
    { \
        static char command[64]; \
        if ((!condition)) return -1; \
        sprintf(command, "SET SELF " #ssip_name " %d", val); \
        return spd_execute_command(connection, command); \
    } \

#define SPD_SET_COMMAND_STR(param, ssip_name) \
    int \
    spd_set_ ## param (int connection, const char *str) \
    { \
        char *command; \
        int ret; \
        if (str == NULL) return -1; \
        command = g_strdup_printf("SET SELF " #param " %s", str); \
        ret = spd_execute_command(connection, command); \
        spd__free(command); \
        return ret; \
    } \

SPD_SET_COMMAND_INT(rate, RATE, ((val >= -100) && (val <= +100)) );
SPD_SET_COMMAND_INT(pitch, PITCH, ((val >= -100) && (val <= +100)) );
SPD_SET_COMMAND_INT(volume, VOLUME, ((val >= -100) && (val <= +100)) );

SPD_SET_COMMAND_STR(voice_type, VOICE);

#undef SPD_SET_COMMAND_INT
#undef SPD_SET_COMMAND_STR

int
spd_execute_command(int connection, char* command)
{
    char *buf;
    char *ret;
    int r;
    
    buf = g_strdup_printf("%s\r\n", command);
    ret = spd_send_data(connection, buf, SPD_WAIT_REPLY);
    spd__free(buf);
    
    r = spd__ret_ok(ret);
    spd__free(ret);

    if (!r) return -1;
    else return 0;
}

#define SPD_MAX_REPLY_LENGTH 1024
char*
spd_send_data(int fd, const char *message, int wfr)
{
    char *reply;
    int bytes;

    reply = malloc(sizeof(char) * SPD_MAX_REPLY_LENGTH);
   
    /* write message to the socket */
    write(fd, message, strlen(message));
    DBG(">> : |%s|", message);

    /* read reply to the buffer */
    if (wfr){
        bytes = read(fd, reply, SPD_MAX_REPLY_LENGTH);
        if (bytes == -1){
            DBG("Error: Can't read reply, broken socket.");
            return NULL;
        }
        /* print server reply to as a string */
        reply[bytes] = 0; 
        DBG("<< : |%s|\n", reply);
    }else{
        DBG("<< : no reply expected");
        return "NO REPLY";
    } 

    return reply;
}


/* --------------------- Internal functions ------------------------- */

int
spd__set_priority(int connection, SPDPriority priority)
{
    static char p_name[16];
    static char command[64];

    switch(priority){
    case SPD_IMPORTANT: strcpy(p_name, "IMPORTANT"); break;
    case SPD_MESSAGE: strcpy(p_name, "MESSAGE"); break;
    case SPD_TEXT: strcpy(p_name, "TEXT"); break;
    case SPD_NOTIFICATION: strcpy(p_name, "NOTIFICATION"); break;
    case SPD_PROGRESS: strcpy(p_name, "PROGRESS"); break;
    default: 
        DBG("Error: Can't set priority! Incorrect value.");
        return -1;
    }
		 
    sprintf(command, "SET SELF PRIORITY %s", p_name);
    return spd_execute_command(connection, command);
}

int
spd__ret_ok(char *reply)
{
	int err;

	err = spd__get_err_code(reply);
		
	if ((err>=100) && (err<300)) return 1;
	if (err>=300) return 0;

	SPD_FATAL("Internal error during communication.");
}

int
spd__get_err_code(char *reply)
{
    char err_code[4];
    int err;
	
    if (reply == NULL) return -1;
    DBG("spd_send_data:	reply: %s\n", reply);

    err_code[0] = reply[0];	err_code[1] = reply[1];
    err_code[2] = reply[2];	err_code[3] = '\0';

    DBG("spd__ret_ok: err_code:	|%s|\n", err_code);
   
    if(spd__isanum(err_code)){
        err = atoi(err_code);
    }else{
        DBG("spd__ret_ok: not a number\n");
        return -1;	
    }

    return err;
}

/* spd__isanum() tests if the given string is a number,
 *  returns 1 if yes, 0 otherwise. */
int
spd__isanum(char *str)
{
    int i;
    if (str == NULL) return 0;
    for(i=0;i<=strlen(str)-1;i++){
        if (!isdigit(str[i]))   return 0;
    }
    return 1;
}

void
spd__free(void *ptr)
{
    if (ptr != NULL){
        free(ptr);
        ptr = NULL;
    }
}

char*
spd__escape_dot(const char *text)
{
    char *seq;
    GString *ntext;
    char *p;
    char *otext;
    char *ret = NULL;
    int len;

    if (text == NULL) return NULL;

    DBG("Incomming text to escaping: |%s|", text);

    p = (char*) text;
    otext = p;

    ntext = g_string_new("");

    len = strlen(text);
    if (len == 1){
        if (!strcmp(text, ".")){
            g_string_append(ntext, "..");
            p += 1;
        }
    }
    else if (len >= 3){
        if ((p[0] == '.') && (p[1] == '\r') && (p[2] == '\n')){
            g_string_append(ntext, "..\r\n");
            p += 3;
        }
    }

    //    DBG("Altering text (I): |%s|", ntext->str);

    while ((seq = strstr(p, "\r\n.\r\n"))){
        assert(seq>p);
        g_string_append_len(ntext, p, seq-p);
        g_string_append(ntext, "\r\n..\r\n");
        p = seq+5;
    }

    //   DBG("Altering text (II): |%s|", ntext->str);    

    len = strlen(p);
    if (len >= 3){
        if ((p[len-3] == '\r') && (p[len-2] == '\n')
            && (p[len-1] == '.')){
            g_string_append(ntext, p);
            g_string_append(ntext, ".");
            p += len;
            //  DBG("Altering text (II-b): |%s|", ntext->str);    
        }
    }

    if (p == otext){
        DBG("No escaping needed.");
        g_string_free(ntext, 1);
        return NULL;
    }else{
        g_string_append(ntext, p);
        ret = ntext->str;
        g_string_free(ntext, 0);
    }

    DBG("Altered text after escaping: |%s|", ret);

    return ret;
}



libspeechd.h

/*
 * libspeechd.h - functions for acces to Speech Dispatcher functions (header)
 *
 * Copyright (C) 2005 Brailcom, o.p.s.
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This software is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
 * License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this package; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 * $Id: libspeechd.h,v 1.1.1.1 2005-12-16 16:29:50 hanke Exp $
 */


#include <stdio.h>

#define SPEECHD_DEFAULT_PORT 6560

typedef enum 
{                  /* Type of voice */
    NO_VOICE = 0,
    MALE1 = 1,
    MALE2 = 2,
    MALE3 = 3,
    FEMALE1 = 4,
    FEMALE2 = 5,
    FEMALE3 = 6,
    CHILD_MALE = 7,
    CHILD_FEMALE = 8
}EVoiceType;

typedef enum
{
    SORT_BY_TIME = 0,
    SORT_BY_ALPHABET = 1
}ESort;

typedef enum
{
    MSGTYPE_TEXT = 0,
    MSGTYPE_SOUND_ICON = 1,
    MSGTYPE_CHAR = 2,
    MSGTYPE_KEY = 3,
    MSGTYPE_SPELL = 99
}EMessageType;

typedef enum
{
    RECOGN_NONE = 0,
    RECOGN_SPELL = 1,
    RECOGN_ICON = 2
}ECapLetRecogn;

typedef enum
{
    PUNCT_NONE = 0,
    PUNCT_ALL = 1,
    PUNCT_SOME = 2
}EPunctMode;

typedef enum
{
    SPELLING_OFF = 0,
    SPELLING_ON = 1
}ESpellMode;


/* Debugging */
FILE* spd_debug;
void SPD_DBG(char *format, ...);

/* Unless there is an fatal error, it doesn't print anything */
#define SPD_FATAL(msg) { printf("Fatal error (libspeechd) [%s:%d]:"msg, __FILE__, __LINE__); exit(EXIT_FAILURE); }

/* Arguments for spd_send_data() */
#define SPD_WAIT_REPLY 1              /* Wait for reply */
#define SPD_NO_REPLY 0               /* No reply requested */


/* --------------------- Public data types ------------------------ */

typedef enum{
    SPD_PUNCT_ALL = 0,
    SPD_PUNCT_NONE = 1,
    SPD_PUNCT_SOME = 2
}SPDPunctuation;

typedef enum{
    SPD_CAP_NONE = 0,
    SPD_CAP_SPELL = 1,
    SPD_CAP_ICON = 2
}SPDCapitalLetters;

typedef enum{
    SPD_SPELL_OFF = 0,
    SPD_SPELL_ON = 1
}SPDSpelling;

typedef enum{
    SPD_MALE1 = 1,
    SPD_MALE2 = 2,
    SPD_MALE3 = 3,
    SPD_FEMALE1 = 4,
    SPD_FEMALE2 = 5,
    SPD_FEMALE3 = 6,
    SPD_CHILD_MALE = 7,
    SPD_CHILD_FEMALE = 8
}SPDVoiceType;

typedef enum{
    SPD_IMPORTANT = 1,
    SPD_MESSAGE = 2,
    SPD_TEXT = 3,
    SPD_NOTIFICATION = 4,
    SPD_PROGRESS = 5
}SPDPriority;

/* -------------- Public functions --------------------------*/

/* Openning and closing Speech Dispatcher connection */
int spd_open(const char* client_name, const char* connection_name, const char *user_name);
void spd_close(int connection);

/* Speaking */
int spd_say(int connection, SPDPriority priority, const char* text);

/* Speech flow */
int spd_stop(int connection);

int spd_cancel(int connection);

int spd_pause(int connection);

int spd_resume(int connection);

/* Sound icons */
int spd_sound_icon(int connection, SPDPriority priority, const char *icon_name);

/* Setting parameters */
int spd_set_voice_type(int connection, const char *voice);

int spd_set_rate(int connection, signed int rate);

int spd_set_pitch(int connection, signed int pitch);

int spd_set_volume(int connection, signed int volume);

/* Direct SSIP communication */
int spd_execute_command(int connection, char* command);
char* spd_send_data(int fd, const char *message, int wfr);

/* --------------  Private functions  ------------------------*/

int spd__set_priority(int connection, SPDPriority priority);

char* spd__escape_dot(const char *otext);

int spd__isanum(char* str);		
void spd__free(void *ptr);
int spd__ret_ok(char *reply);
int spd__get_err_code(char *reply);


speechdspeaker.h
/*
 * GNOME Speech - Speech services for the GNOME desktop
 *
 * Copyright 2002 Sun Microsystems Inc.
 * Copyright 2005 Brailcom, o.p.s.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 * speechdspeaker.h: Definition of the SpeechdSynthesisDriver
 *                            object-- a GNOME Speech driver for the Speechd
 *                            Speech Synthesis System (implementation in
 *                            speechdsynthesisdriver.c)
 *
 */

#ifndef __SPEECHD_SPEAKER_H_
#define __SPEECHD_SPEAKER_H_

#include <bonobo/bonobo-object.h>
#include <gnome-speech/gnome-speech.h>

#define SPEECHD_SPEAKER_TYPE        (speechd_speaker_get_type ())
#define SPEECHD_SPEAKER(o)          (G_TYPE_CHECK_INSTANCE_CAST ((o), SPEECHD_SPEAKER_TYPE, SpeechdSpeaker))
#define SPEECHD_SPEAKER_CLASS(k)    (G_TYPE_CHECK_CLASS_CAST((k), SPEECHD_SPEAKER_TYPE, SpeechdSpeakerClass))
#define IS_SPEECHDSPEAKER(o)       (G_TYPE_CHECK_INSTANCE_TYPE ((o), SPEECHD_SPEAKER_TYPE))
#define IS_SPEECHD_SPEAKER_CLASS(k) (G_TYPE_CHECK_CLASS_TYPE ((k), SPEECHD_SPEAKER_TYPE))

typedef struct {
	Speaker parent;

        int conn;

	gchar *voice;        
} SpeechdSpeaker;

typedef struct {
	SpeakerClass parent_class;
} SpeechdSpeakerClass;

GType
speechd_speaker_get_type   (void);

SpeechdSpeaker *
speechd_speaker_new (GObject *driver,
		      const GNOME_Speech_VoiceInfo *spec);
 
#endif /* __SPEECHD_SPEAKER_H_ */


speechdsynthesisdriver.c
/*
 *
 * GNOME Speech - Speech services for the GNOME desktop
 *
 * Copyright 2002 Sun Microsystems.
 * Copyright 2005 Brailcom, o.p.s.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 *
 * speechdsynthesisdriver.c: Implementation of the SpeechDispatcherSynthesisDriver
 *                            object-- a GNOME Speech driver for the Festival
 *                            Speech Synthesis System
 *
 */

#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <signal.h>
#include <unistd.h>
#include <stdlib.h>
#include <libbonobo.h>
#include <netdb.h>
#include "speechdsynthesisdriver.h"
#include "speechdspeaker.h"
#include "libspeechd.h"

/* Put a message into the logfile (stderr) */
#define DBG(arg...) \
{ \
    fprintf(stderr," SpeechdSynthesisDriver: "); \
    fprintf(stderr,arg); \
    fprintf(stderr,"\n"); \
    fflush(stderr); \
}


static GObjectClass 	*parent_class;
static gboolean 	speechd_server_got_voices = FALSE;
static GSList		*driver_list = NULL;
static GSList		*voices_list = NULL;

static void 	speechd_add_driver	    	(SpeechdSynthesisDriver *d);
static void 	speechd_remove_driver 		(SpeechdSynthesisDriver *d);
static gboolean speechd_is_any_driver_alive 	(void);


static void 
speechd_add_driver (SpeechdSynthesisDriver *d)
{
	driver_list = g_slist_append (driver_list, d);
}


static void 
speechd_remove_driver (SpeechdSynthesisDriver *d)
{
	driver_list = g_slist_remove (driver_list, d);
}



static gboolean 
speechd_is_any_driver_alive ()
{
	return driver_list != NULL;    
}


static SpeechdSynthesisDriver *
speechd_synthesis_driver_from_servant (PortableServer_Servant *servant)
{
	return SPEECHD_SYNTHESIS_DRIVER(bonobo_object_from_servant(servant));
}


static CORBA_string
speechd__get_driverName (PortableServer_Servant servant,
			  CORBA_Environment * ev)
{
	return CORBA_string_dup ("Speech Dispatcher GNOME Speech Driver");  
}




static CORBA_string
speechd__get_synthesizerName (PortableServer_Servant servant,
			       CORBA_Environment * ev)
{
	return CORBA_string_dup ("Speech Dispatcher");
}



static CORBA_string
speechd__get_driverVersion (PortableServer_Servant aservant,
			     CORBA_Environment * ev)
{
	return CORBA_string_dup ("0.1");
}



static CORBA_string
speechd__get_synthesizerVersion (PortableServer_Servant servant,
				  CORBA_Environment * ev)
{
	SpeechdSynthesisDriver *d = speechd_synthesis_driver_from_servant (servant);
	return CORBA_string_dup (d->version);
}

static void
voice_list_free (GSList *l)
{
	g_assert (l);
	GSList *tmp = l;
	
	while (tmp)
	{
    	    CORBA_free (tmp->data);
	    tmp = tmp->next;
	}
	g_slist_free (l);
}



static GSList *
get_voice_list_with_spec (GSList *l,
			  const GNOME_Speech_VoiceInfo *info)
{
	GSList *tmp, *new;
	GNOME_Speech_VoiceInfo *i, *new_info;

	new = NULL;
	tmp = l;
	while (tmp) {
		i = (GNOME_Speech_VoiceInfo *) tmp->data;
		if ((strlen(info->name) ? strcmp (i->name, info->name) : 1) && 
		    (strlen(info->language) ? strcmp (i->language, info->language) : 1) &&
		    (info->gender == i->gender)) {
			new_info = GNOME_Speech_VoiceInfo__alloc ();
            		new_info->language = CORBA_string_dup(info->language);
            		new_info->name = CORBA_string_dup (info->name);
            		new_info->gender = info->gender;
			new = g_slist_prepend (new, new_info);
		}
		tmp = tmp->next;
	}
	if (new == NULL) return NULL;

	new = g_slist_reverse (new);
	return new;
}


static GNOME_Speech_VoiceInfoList *
voice_info_list_from_voice_list (GSList *l)
{
	int i = 0;
	GNOME_Speech_VoiceInfoList *rv = GNOME_Speech_VoiceInfoList__alloc ();
  
	if (!l) {
		rv->_length = rv->_maximum = 0;
		return rv ;
	}

	rv->_length = rv->_maximum = g_slist_length (l);
	rv->_buffer = GNOME_Speech_VoiceInfoList_allocbuf (rv->_length);

	while (l) {
		GNOME_Speech_VoiceInfo *info =
			(GNOME_Speech_VoiceInfo *) l->data;
		rv->_buffer[i].name = CORBA_string_dup (info->name);
		rv->_buffer[i].gender = info->gender;
		rv->_buffer[i].language = CORBA_string_dup(info->language);
		i++;
		l = l->next;
	}
	return rv;
}



static GNOME_Speech_VoiceInfoList *
speechd_getVoices (PortableServer_Servant servant,
		    const GNOME_Speech_VoiceInfo *voice_spec,
		    CORBA_Environment *ev)
{
	GNOME_Speech_VoiceInfoList *rv;
	GSList *l;

	DBG("speechd_getVoices");

	l = get_voice_list_with_spec (voices_list, voice_spec);
	if (l == NULL) return NULL;

	rv = voice_info_list_from_voice_list (l);
	voice_list_free (l);
	return rv;
}



static GNOME_Speech_VoiceInfoList *
speechd_getAllVoices (PortableServer_Servant servant,
		       CORBA_Environment *ev)
{
	GNOME_Speech_VoiceInfoList *rv;

	DBG("speechd_getAllVoices");

	rv = voice_info_list_from_voice_list (voices_list);
	return rv;
}



static GNOME_Speech_Speaker
speechd_createSpeaker (PortableServer_Servant servant,
			const GNOME_Speech_VoiceInfo *voice_spec,
			CORBA_Environment *ev)
{
	SpeechdSynthesisDriver *d = speechd_synthesis_driver_from_servant (servant);
	SpeechdSpeaker *s;
	GNOME_Speech_VoiceInfo *info;
	GSList *l;

	DBG("speechd_createSpeaker");
	
	l = get_voice_list_with_spec (voices_list, voice_spec);
	
	if (l)
		info = l->data;
	else
		info = NULL;

	s = speechd_speaker_new (G_OBJECT(d), info);
	/* FIXME: shound't we free l here? */
	return CORBA_Object_duplicate(bonobo_object_corba_objref (BONOBO_OBJECT(s)), ev);
}


#define ADD_VOICE(voice_name, voicegen) \
	new_info = GNOME_Speech_VoiceInfo__alloc (); \
	new_info->name = CORBA_string_dup (voice_name); \
	new_info->language = CORBA_string_dup ("any"); \
	new_info->gender = GNOME_Speech_gender_##voicegen; \
	voices_list = g_slist_append (voices_list, new_info);

static CORBA_boolean
speechd_driverInit (PortableServer_Servant servant,
		     CORBA_Environment *ev)
{
        GNOME_Speech_VoiceInfo *new_info;
	
        SpeechdSynthesisDriver *d = speechd_synthesis_driver_from_servant (servant);

	DBG("speechd_driverInit");

	if (d->initialized) return TRUE;

	d->version = strdup("unknown");
	if (!d->version) return FALSE;

	DBG("adding voices to list");
	ADD_VOICE("MALE1", male);
	ADD_VOICE("MALE2", male);
	ADD_VOICE("MALE3", male);
	ADD_VOICE("FEMALE1", female);
	ADD_VOICE("FEMALE2", female);
	ADD_VOICE("FEMALE3", female);
	ADD_VOICE("CHILD_MALE", male);
	ADD_VOICE("CHILD_FEMALE", female);

	speechd_server_got_voices = TRUE;	
	d->initialized = TRUE;

	DBG("driver initialization finished");
	return TRUE;
}


static CORBA_boolean
speechd_isInitialized (PortableServer_Servant servant,
			CORBA_Environment *ev)
{
	SpeechdSynthesisDriver *d = speechd_synthesis_driver_from_servant (servant);

	return d->initialized;
}


static void
speechd_synthesis_driver_init (SpeechdSynthesisDriver *d)
{
        d->version = NULL;
        d->last_speaker = NULL;
        d->initialized = FALSE;
        d->sock = -1;
	d->pipe = -1;
        d->channel_sock = NULL;
        d->channel_pipe = NULL;	
        d->crt_clbs = NULL;
        d->is_shutting_up = FALSE;
	d->is_speaking = FALSE;
        d->is_querying = FALSE;
        d->list = NULL;
        d->crt_id = 0;
        d->queue_length = 0;	
}


static void
speechd_synthesis_driver_finalize (GObject *obj)
{
        SpeechdSynthesisDriver *d = SPEECHD_SYNTHESIS_DRIVER (obj);
    	
	if (parent_class->finalize)
    	    parent_class->finalize (obj);
		
	speechd_remove_driver (d);
	
        if (!speechd_is_any_driver_alive ())
	    bonobo_main_quit ();
}


static void
speechd_synthesis_driver_class_init (SpeechdSynthesisDriverClass *klass)
{
	GObjectClass *object_class = G_OBJECT_CLASS(klass);
	parent_class = g_type_class_peek_parent (klass);

	object_class->finalize = speechd_synthesis_driver_finalize;
	
	/* Initialize epv table */

	klass->epv._get_driverName = speechd__get_driverName;
	klass->epv._get_driverVersion = speechd__get_driverVersion;
	klass->epv._get_synthesizerName = speechd__get_synthesizerName;
	klass->epv._get_synthesizerVersion = speechd__get_synthesizerVersion;
	klass->epv.getVoices = speechd_getVoices;
	klass->epv.getAllVoices = speechd_getAllVoices;
	klass->epv.createSpeaker = speechd_createSpeaker;
	klass->epv.driverInit = speechd_driverInit;
	klass->epv.isInitialized = speechd_isInitialized;
}


BONOBO_TYPE_FUNC_FULL (SpeechdSynthesisDriver,
		       GNOME_Speech_SynthesisDriver,
		       bonobo_object_get_type (),
		       speechd_synthesis_driver);


SpeechdSynthesisDriver * 
speechd_synthesis_driver_new (void)
{
        SpeechdSynthesisDriver *driver;
	
	driver = g_object_new (SPEECHD_SYNTHESIS_DRIVER_TYPE, NULL);
	speechd_add_driver (driver);
    
	return driver;
}


int
main (int  argc,
      char **argv)
{
        SpeechdSynthesisDriver *driver;
	char 	*obj_id;
	int 	ret;

        driver_list = NULL;

	if (!bonobo_init (&argc, argv))
        {
    	    g_error ("Could not initialize Bonobo Activation / Bonobo");
	}

        obj_id = "OAFIID:GNOME_Speech_SynthesisDriver_Speech_Dispatcher:proto0.3";

	driver = speechd_synthesis_driver_new ();

	if (!driver)
    	    g_error ("Error creating speech synthesis driver object.\n");

        ret = bonobo_activation_active_server_register (obj_id,
					                bonobo_object_corba_objref (bonobo_object (driver)));

        if (ret != Bonobo_ACTIVATION_REG_SUCCESS)
	    g_error ("Error registering speech synthesis driver.\n");
	else
	    bonobo_main ();

        g_assert (driver_list == NULL);
	bonobo_debug_shutdown ();	

        return 0;
}
